### DISCUSSION TITLE
How to apply `emotion` style to ErrorBoundary

### DISCUSSION THREAD

**Discussion by Kazuhiro-Mimaki** [remix-run/remix#6543] [Q&A]
I use remix with emotion as style framework.

If I throw error in action function, remix redirect to ErrorBoundary(in root).
However, style is not applied in the error page.

Why emotion is not applied to redirected error page and how to?

Following code is example for `root.tsx`.
If `throw new Error()` in loader/action function in routes, ErrorBoundary catch the error and display the body of ErrorBoundary, but emotion style is not applied.

root.tsx
```tsx
import { css } from '@emotion/react';
import type { LinksFunction, MetaFunction } from '@remix-run/node';
import { Links, LiveReload, Meta, Outlet, Scripts, ScrollRestoration, useRouteError } from '@remix-run/react';

export const meta: MetaFunction = () => ({
  charset: 'utf-8',
  title: 'New Remix App',
  viewport: 'width=device-width,initial-scale=1',
});

export default function App() {
  return (
    <html lang='en'>
      <head>
        <Meta />
        <Links />
      </head>
      <body>
        <Outlet />
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}

export function ErrorBoundary() {
  return (
    <html lang='en'>
      <head>
        <Meta />
        <Links />
      </head>
      <body>
        <div css={style}>
          <div>hoge</div>
          <div>fuga</div>
        </div>
      </body>
    </html>
  );
}

const style = css({
  display: 'flex',
  backgroundColor: 'red',
});
```

**Comment by fernandojbf** [comment]
You can use a layout route and put the ErrorBoundary there. That way the root will render the ErrorBoundary as Outlet. 

https://remix.run/docs/en/main/guides/routing#pathless-layout-routes

example:
```
/routes/_app.tsx (layout with error boundary)
/routes/_app._index.tsx (index page)
/routes/_app....tsx (rest of the app)
root.tsx
```

ErrorBoundary from root.tsx does not render the Root element. is the "ultimate" fallback (if root fails, you need something than is deterministic and do not depend from others.

**Reply by Kazuhiro-Mimaki** [reply]
Thank you! I can by pathless layout routes!
However, I wonder why root error boundary is not styled with emotion.

> ErrorBoundary from root.tsx does not render the Root element. is the "ultimate" fallback (if root fails, you need something than is deterministic and do not depend from others.

I'm sorry but I cannot understand what you mean.

In docs it seems that root error boundary can be useful, but isn't it?
https://remix.run/docs/en/main/guides/errors#root-error-boundary

**Reply by fernandojbf** [reply]
Sorry I missed the part that you were using emotion in the boundary ( i thought was in the root). 

The ErrorBoundary is useful for many things. 

Did you prepared remix for the SSR styles (https://emotion.sh/docs/ssr)? Looking only to the file you sent, probably your styles are been only added client side. Since you didn't add <Scripts /> in you ErrorBoundary the styles are not injected because JS is not added to the page

If you only make a solution client side, you will have the SSR been render without styles.

Other thing: I do not use emotion from a long time but I believe `css` prop only work with some integration steps (check https://emotion.sh/docs/css-prop). Remix bring some complexity to that due to the impossibility to configure the bundler. So if emotions require bundler integration to add styles you will have some problems xD

I believe emotion works without any integration but will not be full featured

have a look on: https://remix.run/docs/en/1.16.1/guides/styling

**Reply by Kazuhiro-Mimaki** [reply]
Thank you!

I didn't prepare SSR styles.
Remix of current version use `renderToPipeableStream()` in entry.server.tsx, but `renderToPipeableStream()` is not mentioned in the page (https://emotion.sh/docs/ssr).

How to use emotion with `renderToPipeableStream()`? Do you have any idea?

I tried to implement like following code, but not work successfully.
```tsx
function handleBrowserRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  return new Promise((resolve, reject) => {
    let shellRendered = false;
    const emotionCache = createEmotionCache();

    const { pipe, abort } = renderToPipeableStream(
      <EmotionCacheProvider value={emotionCache}>
        <RemixServer
          context={remixContext}
          url={request.url}
          abortDelay={ABORT_DELAY}
        />
      </EmotionCacheProvider>,
      {
        onShellReady() {
          shellRendered = true;
          const body = new PassThrough();

          const emotionServer = createEmotionServer(emotionCache);

          const bodyWithStyles = emotionServer.renderStylesToNodeStream();
          body.pipe(bodyWithStyles);

          responseHeaders.set("Content-Type", "text/html");

          resolve(
            new Response(body, {
              headers: responseHeaders,
              status: responseStatusCode,
            })
          );

          pipe(body);
        },
        onShellError(error: unknown) {
          reject(error);
        },
        onError(error: unknown) {
          responseStatusCode = 500;
          // Log streaming rendering errors from inside the shell.  Don't log
          // errors encountered during initial shell rendering since they'll
          // reject and get logged in handleDocumentRequest.
          if (shellRendered) {
            console.error(error);
          }
        },
      }
    );

    setTimeout(abort, ABORT_DELAY);
  });
}
```

**Reply by fernandojbf** [reply]
Like I mentioned, I did not used emotion from a long time and I did not try to integrate with remix :( But that integration seems odd to me. You are passing body as argument of renderStylesToNodeStream, but this function does not receive any argument https://github.com/emotion-js/emotion/blob/fef1ea086fbec998caaa016a5ecdace4bbb9d749/packages/server/src/create-instance/stream.js#L8. 

I believe you will need to continue trying and debugging. 

Good luck :)

**Reply by Kazuhiro-Mimaki** [reply]
> But that integration seems odd to me. You are passing body as argument of renderStylesToNodeStream, but this function does not receive any argument https://github.com/emotion-js/emotion/blob/fef1ea086fbec998caaa016a5ecdace4bbb9d749/packages/server/src/create-instance/stream.js#L8.

Oh, that is my mistake...
Thank you so much.

**Reply by kiliman** [reply]
To clarify, when Remix encounters an error, it will render the nearest `<ErrorBoundary>`. This *replaces* the default export. So for your root `<ErrorBoundary>` it needs all the components you would render for the default export. I typically create a `<Document>` wrapper that is used by both the default export and the error boundary.

```ts
// root.tsx

function Document({
  children,
  title,
}: {
  children: React.ReactNode;
  title?: string;
}) {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta
          name="viewport"
          content="width=device-width, initial-scale=1, viewport-fit=cover"
        />
        {title ? <title>{title}</title> : null}
        <Meta />
        <Links />
      </head>
      <body>
        {children}
        <Scripts />
        <ScrollRestoration />
        {process.env.NODE_ENV === 'development' && <LiveReload />}
      </body>
    </html>
  );
}

export default function App() {
  return (
    <Document>
      <Outlet />
    </Document>
  )
}

export function ErrorBoundary() {
  const error = useRouteError();
  let title: string;
  if (isRouteErrorResponse(error)) {
    title = `Error ${error.status} - ${error.statusText}`;
  } else {
    const { message } = error as Error;
    title = `Application Error - ${message}`;
  }

  return (
    <Document title={title}>
      {/* render error here */}
    </Document>
  );
}
```

**Reply by Kazuhiro-Mimaki** [reply]
Thank you! That technique is nice.

> To clarify, when Remix encounters an error, it will render the nearest <ErrorBoundary>. This replaces the default export.

I understand this behavior, but I don't why emotion css is not applied to ErrorBoundary in some case.

For example, link from routeA to routeB, and error is thrown in routeB loader, ErrorBoundary catch the error but doesn't apply emotion css.

RouteA
```tsx
import { Link } from "@remix-run/react";

export default function RouteA() {
  return (
    <div>
      <Link to="/B">click</Link>
    </div>
  );
}
```

RouteB
```tsx
export const loader = () => {
  throw new Error();
};

export default function RouteB() {
  return <div>...</div>;
}
```

Then, root.tsx catch the error by ErrorBoundary. But the emotion css is not applied. If reload the page that is applied. Why?

```tsx
import { css } from "@emotion/react";
import type { MetaFunction } from "@remix-run/node";
import {
  Links,
  LiveReload,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
} from "@remix-run/react";

export const meta: MetaFunction = () => ({
  charset: "utf-8",
  title: "New Remix App",
  viewport: "width=device-width,initial-scale=1",
});

export default function App() {
  return (
    <html lang="en">
      <head>
        <Meta />
        <Links />
      </head>
      <body>
        <Outlet />
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}

export function ErrorBoundary() {
  return (
    <html lang="en">
      <head>
        <Meta />
        <Links />
      </head>
      <body>
        <div css={style}>error</div>
      </body>
    </html>
  );
}

const style = css({
  backgroundColor: "red",
});
```

**Reply by kiliman** [reply]
Have you looked at the _emotion_ example from the Remix repo? Since client-side navigations don't go through _entry.server_, you need to keep track of the emotion styles cache on the client by wrapping `<RemixBrowser>`. This ensures emotion properly updates the styles as new routes are rendered.

https://github.com/remix-run/examples/tree/main/emotion

If you run the codesandbox example, you'll see that the `<ErrorBoundary>` styles are properly rendered on the client (click the link to Jokes Error), as well as the server (refresh the error page... best if you pop out the browser to a new tab).

**Reply by Kazuhiro-Mimaki** [reply]
Thank you!
This example use `renderToString`, however, in the latest remix, I use `renderToPipeableStream` and not work successfully with the emotion example.
Do you have any ideas about use `renderToPipeableStream` with emotion?
